pkg/syncer/controllermanager/controllermanager.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllermanager
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/logging"
+)
+
+const (
+	ControllerNamePrefix = "syncer-controller-manager-"
+)
+
+type InformerSource struct {
+	Subscribe func(id string) <-chan struct{}
+	Informer  func(gvr schema.GroupVersionResource) (informer cache.SharedIndexInformer, known, synced bool)
+}
+
+type Controller interface {
+	Start(ctx context.Context, numThreads int)
+}
+
+type ControllerDefintion struct {
+	RequiredGVRs []schema.GroupVersionResource
+	NumThreads   int
+	Create       func(syncedInformers map[schema.GroupVersionResource]cache.SharedIndexInformer) (Controller, error)
+}
+
+func NewControllerManager(ctx context.Context, suffix string, informerSource InformerSource, controllers map[string]ControllerDefintion) *ControllerManager {
+	controllerManager := ControllerManager{
+		name:                  ControllerNamePrefix + suffix,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerNamePrefix+suffix),
+		informerSource:        informerSource,
+		controllerDefinitions: controllers,
+		startedControllers:    map[string]context.CancelFunc{},
+	}
+
+	apisChanged := informerSource.Subscribe(controllerManager.name)
+
+	logger := klog.FromContext(ctx)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-apisChanged:
+				logger.V(4).Info("got API change notification")
+				controllerManager.queue.Add("resync") // this queue only ever has one key in it, as long as it's constant we are OK
+			}
+		}
+	}()
+
+	return &controllerManager
+}
+
+type ControllerManager struct {
+	name                  string
```

Comment:
why does this object know its name?

Reply:
Because contrary to typical controllers (which it is not), there can be several instances of controller managers.
Typically one per ddsif.
So the name is useful to distinguish controller manager instances in logs, as well as when subscribing to the informer source (passing the instance name as the id)